concurrency/DropletClassMapper.java

```diff
@@ -0,0 +1,221 @@
+     * @throws UnknownDropletClassException If the droplet class is unknown.
+     */
+    public int getDropletClassId(String dropletClassName) throws DropletClassMapperException, UnknownDropletClassException {
+        Integer id = dropletClassNameIdMap.get(dropletClassName);
+        if (id == null) {
+            id = getDropletClassIdFromDataSource(dropletClassName);
+            dropletClassNameIdMap.put(dropletClassName, id);
+            if (id != null) {
+                dropletClassIdNameMap.put(id, dropletClassName);
+            }
+        }
+        return id;
+    }
+
+    /**
+     * Method to get the mapping from the data source, bypassing the cache.
+     * Can be overridden for test purposes.
+     * @param dropletClassName
+     * @return The integral ID for the name.
+     * @throws SQLException
+     */
+    protected int getDropletClassIdFromDataSource(String dropletClassName) throws DropletClassMapperException, UnknownDropletClassException {
+        try {
+            return getDropletDbHelper().getDropletClassIdFromName(dropletClassName);
+        } catch (NoRowsReturnedException e) {
+            throw new UnknownDropletClassException("No droplet-class with name '" + dropletClassName + "' was found.", e);
+        } catch (Exception e) {
+            throw new DropletClassMapperException("Error occurred finding a mapping for droplet-class with name '" + dropletClassName + "'.", e);
+        }
+    }
+
+    /**
+     * Gets the the droplet class value name for a given droplet class ID and droplet class value ID
+     * @param dropletClassId The droplet class ID.
+     * @param dropletClassValueId The droplet class value ID.
+     * @return The name of the droplet class value.
+     * @throws DropletClassMapperException If something went wrong retrieving the mapping.
+     * @throws UnknownDropletClassValueException If the droplet class value is unknown for the given droplet class.
+     */
+    public String getDropletClassValueName(int dropletClassId, int dropletClassValueId) throws DropletClassMapperException, UnknownDropletClassValueException {
+        /* First just get the mapping for the droplet-class-id to ensure that it is valid */
+        getDropletClassName(dropletClassId);
+
+        /* Get the two maps we will need for this */
+        Map<Integer, String> dcvIdNameMap = dropletClassValueIdNameMaps.get(dropletClassId);
+        Map<String, Integer> dcvNameIdMap = dropletClassValueNameIdMaps.get(dropletClassId);
+
+        /* If either is null, make a new one and put it into the super-map */
+        if (dcvIdNameMap == null) {
+            dcvIdNameMap = new ConcurrentHashMap<Integer, String>();
+            dropletClassValueIdNameMaps.put(dropletClassId, dcvIdNameMap);
+        }
+
+        if (dcvNameIdMap == null) {
+            dcvNameIdMap = new ConcurrentHashMap<String, Integer>();
+            dropletClassValueNameIdMaps.put(dropletClassId, dcvNameIdMap);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `get()`, `null` check, and `put()` may not be thread-safe at lines: **117, 125, and 127**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 The thread whose value was overwritten still has a reference to that value and uses that reference at line: **138**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.